rfcs/20190829-tfx-container-component-execution.md

```diff
@@ -0,0 +1,555 @@
+This approach is closer to what we are proposing in the document. However, we
+cannot directly use the operator because:
+
+*   Airflow operator requires to be run inside an Airflow pipeline which is not
+    the case for local and KF runners.
+*   Airflow operator exposes a subset of POD’s API, where we want to expose the
+    full pod spec to the user.
+*   Airflow operator doesn’t provide a reliable way to retry user’s container
+    and recover from transient errors.
+*   Airflow does not support initializing operator inside another operator.
+    Going back to use multiple Airflow operators for a component is a regression
+    now that we have `BaseComponentLauncher` ready.
+
+## Proposed Design
+
+### TLDR
+
+We propose to solve the above problems by the following design.
+
+*   Define container as an executor spec.
+*   Launch container by component launcher in either local docker or k8s pod.
+*   Use platform config to specify platform specific settings like k8s pod
+    config.
+
+The proposed solution has following parts:
+
+*   Extensible `ExecutorSpec` concept which can support container as an
+    executor.
+*   Extensible `BaseComponentLauncher` concept to support pluggable component
+    launchers in tfx runner.
+    *   `DockerComponentLauncher` which launches `ExecutorContainerSpec` in
+        docker environment.
+    *   `KubernetesPodComponentLauncher` which launches `ExecutorContainerSpec`
+        in k8s environment.
+*   Extensible `PlatformConfig` framework.
+    *   `KubernetesPodPlatformConfig` to support k8s pod spec as a config.
+    *   `DockerPlatformConfig` to support docker run configs.
+
+### Architecture
+
+Architecture that allows local container execution.
+
+![TFX local container execution](20190829-tfx-container-component-execution/tfx-local-container-execution.png)
+
+Architecture that allows Kubernetes container execution.
+
+![TFX k8s container execution](20190829-tfx-container-component-execution/tfx-k8s-container-execution.png)
+
+Class diagram that allows container execution
+
+![TFX container execution_classes](20190829-tfx-container-component-execution/tfx-container-execution-classes.png)
+
+### Python DSL experience
+
+In order to use container base component in TFX DSL, user needs to do following
+steps. Step 1 and Step 2 follow the DSL extension proposed by the other RFC
+(https://github.com/tensorflow/community/pull/146).
+
+#### Step 1: Defines container based component by `ExecutorContainerSpec`
+
+```python
+class MyContainerBasedExampleGen(BaseComponent):
+
+  SPEC_CLASS = types.make_spec_class(
+    inputs={
+      "raw_input": ChannelParameter(type=standard_artifacts.ExternalArtifact),
+    }
+    outputs={
+      "examples": ChannelParameter(type=standard_artifacts.Examples),
+    }
+    parameters={
+      "num_csv_columns": ExecutionParameter(type=int),
+    }
+  )
+
+  EXECUTOR_SPEC = ExecutorContainerSpec(
```

Comment:
(similar to previous comment):  Is ExecutorContainerSpec() specific to Docker?  Or, will it be correct for any OCI-compatible container solution?  If specific to Docker, rename it.

Reply:
Yes, It's not specific to Docker.